Add rough draft BSD BPF sniffer API

david415 opened the pull request:
OK... here's the code. I haven't done a smoke test lately but last time I tested it worked on OpenBSD, FreeBSD and NetBSD.

Maybe we should be calling SetBpfBuflen instead of using the default. OpenBSD and NetBSD both default to a BPF bufLen of 32767 but FreeBSD uses 4096... obviously having a larger bufLen should improve performance.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+// +build darwin dragonfly freebsd netbsd openbsd
+
+/*
+ * This rough sketch of a golang API for a Berkeley packet filter sniffer for BSD variants.
```
Please use the same copyright as the rest of gopacket, specifically:

// Copyright 2012 Google, Inc. All rights reserved.
//
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE file in the root of the source
// tree.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+ * The bpf_wordalign function was borrowed... and readFrames was very much inspired by
+ * design I saw in songgao's ether git repository; the rest of the code I wrote myself ;-)
+ *
+ * Author: David Anthony Stainton
```
Instead of your name here, please add yourself to the AUTHORS file in the top-level directory.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+
+import (
+	"fmt"
+	"github.com/google/gopacket"
```
Split up imports into standard library vs. other:

import (
  "fmt"
  "syscall"
  "time"
  "unsafe"

  "github.com/google/gopacket"
  "golang.org/x/sys/unix"
)

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+
+const wordSize = int(unsafe.Sizeof(uintptr(0)))
+
+func bpf_wordalign(x int) int {
```
please use camelCase for all functions and vars.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+	Timestamp time.Time
+}
+
+type BpfSniffer struct {
```
capitalize acronyms:  BPFSniffer

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+
+// Init is used to initialize a BPF device for promiscuous sniffing.
+// It also starts a goroutine to continuously read frames.
+func (b *BpfSniffer) Init() error {
```
make this part of NewBPFSniffer()?  Or is there other configuration that can be done between NewBPFSniffer() and Init()?

david415 replied:
There are many configuration options the user could possibly care about... for instance zero-copy, filters etc. See here:
- https://golang.org/src/syscall/bpf_bsd.go
- https://www.freebsd.org/cgi/man.cgi?bpf%284%29

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+
+func (b *BpfSniffer) pickBpfDevice() {
+	for i := 0; i < 99; i++ {
+		b.bpfDeviceName = fmt.Sprintf("/dev/bpf%d", i)
```
allow user to pick their device?

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+	b.stopChan <- true
+}
+
+func (b *BpfSniffer) readFrames() {
```
using a readFrames goroutine and readChan channel seems overly complex... instead, how about just calling lines 121-136 in ReadPacketData().

That way, if folks want to do things in a separate goroutine, they can, but if not, they don't have to.  It also makes error propagation from syscall.Read easier (it's just returned as an error from ReadPacketData), and there's no need  for stopChan.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+	return nil
+}
+
+func (b *BpfSniffer) Stop() {
```
Other ones seem to call this Close() instead of Stop()

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,153 @@
+	}
+}
+
+func (b *BpfSniffer) ReadTimedFrame() TimedFrame {
```
I wouldn't export this, since it's just meant to be used internally (I think?)

gconnell commented:
Hey, thanks for this code!

The initial set of comments contains a bunch of stuff, some of it simple style nits (camelCase), some of it annoying legaleze (copyright notice), and some more structural (use of goroutine for readFrames).  Please feel free to push back on anything you don't agree with... this is a conversation, not a mandate :)  And thanks again for taking the time to contribute!

david415 commented:
Thanks for the review! I'll have those changes ready soon.

david415 commented:
I do believe I've made all the changes you requested. I added a Close method as well.

david415 commented:
I've got a little test ethernet program that I use to test this API. After a minor code cleanup I got it to build and run again on FreeBSD. However I wasn't able to adjust the bpf buflen on FreeBSD and I wonder if the other BSDs behave differently.

david415 commented:
adjustable bpf buflen now works.

also... an optional timeout value can be set.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,155 @@
+// +build darwin dragonfly freebsd netbsd openbsd
```
Little nit:  could you move the copyright to the top and put this down below it?

david415 replied:
ok

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,155 @@
+	sniffDeviceName string
+	bpfDeviceName string
+	fd       int
+	stopChan chan bool
```
stopChan and readChan can now be deleted.

david415 replied:
right right

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,155 @@
+	b.readBuffer = make([]byte, b.readBufLen)
+
+
+	err = syscall.SetBpfInterface(b.fd, b.sniffDeviceName)
```
As you've mentioned, there's a bunch of possible options that can be set on this fd... how about this approach:

    opts := &bsdbpf.Options{
      ... set what you'd like ...
    }
    sniffer := bsdbpf.NewBPFSniffer(interfaceName, opts)

Things like "immediate", "timeout", "promisc", etc, could all be in opts.  And we could provide a default, which we use if folks pass in nil:

    default := Options{ReadBufferLen: defaultReadBufferLen, ...}

    func NewBPFSniffer(iface string, opts *Options) {
      if opts == nil { opts = &default }
      ...
    }

david415 replied:
yes i like that approach.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,155 @@
+	return nil
+}
+
+func (b *BPFSniffer) ReadPacketData() ([]byte, gopacket.CaptureInfo, error) {
```
Currently, the use of readBuffer means that ReadPacketData actually doesn't work the way it does for other sniffers.

ReadPacketData should ensure that the returned buffer is owned by the caller and won't be overridden by subsequent calls to ReadPacketData.  For example, the buffer could be passed to a set of goroutines, which may take variable amounts of time to handle the data.  You can implement this by creating a new buffer when you set readBytesComsumed to zero:

    b.readBytesConsumed = 0
    b.readBuffer = make([]byte, ...)

The issue here is that rawFrame, which is returned to the user, references the shared buffer space, so a subsequent call will destroy the user's buffer by writing over its data.

If you'd like to reuse a buffer, you should implement ZeroCopyReadPacketData, see for example http://godoc.org/github.com/google/gopacket/afpacket#TPacket.ZeroCopyReadPacketData.

david415 replied:
Yes you are right about that... good catch. I'll fix that...

BSD BPF does have zero copy capability however I am not interested in implementing that feature at this time.

gconnell replied:
Add a brief comment on ReadPacketData... this works fine:

// ReadPacketData implements gopacket.PacketDataSource.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,155 @@
+	if b.readBytesConsumed >= b.lastReadLen {
+		b.lastReadLen, err = syscall.Read(b.fd, b.readBuffer)
+		if err != nil {
+			return nil, gopacket.CaptureInfo{}, err
```
I'm not sure if syscall.Read guarantees that its first return value will be zero if it returns an error... to be safe here, you may want to reset b.lastReadLen to zero on an error.

david415 commented:
OK... ReadPacketData should be fixed now.

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+type Options struct {
+	bpfDeviceName    string
```
You'll need to export (capitalize) these names, so users of your API can set them.  Also, please add a comment to each detailing what they do

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,169 @@
+	preserveLinkAddr: true,
+}
+
+type BPFSniffer struct {
```
Please add comments to all exported functions and structs, so appropriate godoc can be generated.  Comments should follow http://blog.golang.org/godoc-documenting-go-code format; they should start with the name of the exported object, like so, then describe what it does, what its arguments do, etc:

    // ExportedName foos the bar.
    func ExportedName(...) { }

gconnell commented:
Great (and really fast) work on all this!  I think once we have comments in place, we should be ready to merge this in.

david415 commented:
What do you think of these comments?

gconnell commented on bsdbpf/bsd_bpf_sniffer.go:
```diff
@@ -0,0 +1,199 @@
+
+// Init is used to initialize a BPF device for promiscuous sniffing.
+// It also starts a goroutine to continuously read frames.
+func (b *BPFSniffer) Init() error {
```
Forgot if I mentioned this earlier, but would it make sense to put this Init logic inside of NewBPFSniffer?

    func NewBPFSniffer(...) (*BPFSniffer, error) {
    }

That way we could avoid people forgetting to call init after New.

david415 replied:
oh i overlooked this comment earlier...

gconnell commented:
Nice!  The comments about option fields are particularly good.  I'm happy to merge whenever you think we're ready to go, just let me know.

david415 commented:
Would it be OK if we wrote a few notes about tuning the Options for different behavior and performance? I think it would be nice to link back to a BSD manpage about BPF. The reason I suggest this is because it may not be obvious that the "Immediate" flag if set to false will cause less syscall context switching to occur for N number of packets... with the trade off being slightly "reactive"; the read syscall blocks until the buffer is filled or timeout is reached. Certain applications will achieve "smoother" performance by tuning the buffer length, adjusting the timeout and setting Immediate flag to false.

david415 commented:
I have not yet tested on dragonfly and darwin... however it is very likely that this API will works on all BSD systems that the official sys package has a BPF header struct for. It annoys me greatly that these are variable width depending on which BSD system and architecture it is. I think also if darwin and freebsd default to a bufLen of 4096 that this might possibly clip packets if larger ones are transmitted. Therefore although we should set Immediate to true as a default option because this is probably the behavior that most developers expect we should perhaps set the default bufLen to some non-zero value so that the default Options results in the same size bufLen on all BSD systems.

david415 commented:
Good thing we use a largish default buffer size...

Haha.. the default size of bufLen in freebsd and darwin being 4096 bytes is in fact totally broken for the context of say a laptop with a wifi card and several other link layers:
https://en.wikipedia.org/wiki/Maximum_transmission_unit#Table_of_MTUs_of_common_media

What I'd like to do is provide an additional method that can be used to compute the approximate largest valid BPF buffer length. That is, a simple function with a for loop and an exponential step...

david415 commented:
Yeah our current default bufLen is sensible... whereas max possible bufLen just doesn't make sense. I just now updated the comments and added a GetReadBufLen method. I think this is a good start and if you are OK with additional changes to the comments then we should merge it.

david415 commented:
oh yes i think we should call Init from the constructor...

david415 commented:
OK i moved the initialization to NEWBPFSniffer. During that change I got rid of the calls to panic and incorrect/old comments. Unless you can spot more things to change then I agree we should merge it. I think future additions to this API can easily be backwards compatible...
